File: py/test/formak/ExtendedKalmanFilter_test.py
import warnings
from datetime import datetime, timedelta

import numpy as np
import pytest
from hypothesis import given, reject, settings
from hypothesis.strategies import floats
from numpy.testing import assert_almost_equal
from scipy.stats import multivariate_normal

from formak import python, ui

warnings.filterwarnings("error")


@pytest.mark.skip("Test marked as flaky, passing and failing on the same input")
@given(floats(), floats(), floats())
@settings(deadline=timedelta(seconds=2))
def test_EKF_process_property(state_x, state_y, control_a):
    config = {}
    dt = 0.1

    ui_Model = ui.Model(
        ui.Symbol("dt"),
        set(ui.symbols(["x", "y"])),
        set(ui.symbols(["a"])),
        {
            ui.Symbol("x"): "x + y * dt",
            ui.Symbol("y"): "y + a * dt",
        },
    )
    ekf = python.compile_ekf(
        state_model=ui_Model,
        process_noise={ui.Symbol("a"): 1.0},
        sensor_models={},
        sensor_noises={},
        config=config,
    )

    control_vector = np.array([[control_a]])
    covariance = np.eye(2)
    state_vector = np.array([[state_x, state_y]]).transpose()

    if not np.isfinite(state_vector).all() or not np.isfinite(control_vector).all():
        # reject infinite / NaN inputs
        reject()
    if (np.abs(state_vector) > 1e100).any() or (np.abs(control_vector) > 1e100).any():
        # reject poorly sized inputs
        reject()

    next_state, next_cov = ekf.process_model(
        dt=dt, state=state_vector, covariance=covariance, control=control_vector
    )

    for i, key in enumerate(ekf.state_model.arglist[1 : ekf.state_size + 1]):
        python_version = next_state[i]

        subs_args = [
            (symbol, float(val))
            for (symbol, val) in zip(
                ekf.state_model.arglist,
                [dt] + list(state_vector) + list(control_vector),
            )
        ]
        symbolic_version = ui_Model.state_model[key].subs(subs_args)

        assert_almost_equal(python_version, symbolic_version)

    starting_central_probability = multivariate_normal(cov=covariance).pdf(
        np.zeros_like(state_vector).transpose()
    )
    try:
        ending_central_probability = multivariate_normal(cov=next_cov).pdf(
            np.zeros_like(state_vector).transpose()
        )
    except np.linalg.LinAlgError:
        # reject inputs leading to poorly conditions covariances
        reject()

    try:
        assert ending_central_probability < starting_central_probability
    except AssertionError:
        print("Starting at {:f}".format(starting_central_probability))
        print(covariance)
        print("Ending at {:f}".format(ending_central_probability))
        print(next_cov)
        raise


@pytest.mark.skip(reason="Flaky with different execution times. Needs debugging")
@given(floats(), floats(), floats())
def test_EKF_sensor_property(x, y, a):
    config = {}

    ui_Model = ui.Model(
        ui.Symbol("dt"),
        set(ui.symbols(["x", "y"])),
        set(ui.symbols(["a"])),
        {
            ui.Symbol("x"): ui.Symbol("x") * ui.Symbol("y") + ui.Symbol("x"),
            ui.Symbol("y"): "y + a * dt",
        },
    )
    ekf = python.compile_ekf(
        state_model=ui_Model,
        process_noise={ui.Symbol("a"): 1.0},
        sensor_models={"simple": {ui.Symbol("x"): ui.Symbol("x")}},
        sensor_noises={"simple": np.eye(1)},
        config=config,
    )

    control_vector = np.array([[a]])
    covariance = np.eye(2)
    state_vector = np.array([[x, y]]).transpose()

    if not np.isfinite(state_vector).all() or not np.isfinite(control_vector).all():
        # reject infinite / NaN inputs
        reject()
    if (np.abs(state_vector) > 1e100).any() or (np.abs(control_vector) > 1e100).any():
        # reject poorly sized inputs
        reject()

    next_state, next_cov = ekf.sensor_model(
        "simple", state=state_vector, covariance=covariance, sensor_reading=np.eye(1)
    )
    first_innovation = ekf.innovations["simple"]

    try:
        starting_central_probability = multivariate_normal(cov=covariance).pdf(
            np.zeros_like(state_vector).transpose()
        )
        ending_central_probability = multivariate_normal(cov=next_cov).pdf(
            np.zeros_like(state_vector).transpose()
        )
    except np.linalg.LinAlgError:
        print("starting cov")
        print(covariance)
        print("next_cov")
        print(next_cov)
        raise

    try:
        # more confident in state estimate after sensor update
        assert ending_central_probability > starting_central_probability
    except AssertionError:
        print("starting pdf")
        print(starting_central_probability)
        print("ending pdf")
        print(ending_central_probability)
        raise

    ekf.sensor_model(
        "simple", state=next_state, covariance=next_cov, sensor_reading=np.eye(1)
    )
    second_innovation = ekf.innovations["simple"]

    # state moves towards reading after a sensor update
    assert (
        np.linalg.norm(first_innovation) > np.linalg.norm(second_innovation)
        or np.linalg.norm(first_innovation) == 0.0
    )


@pytest.mark.skip("Flaky due to assertion on test timing at the end")
def test_EKF_sensor_property_failing_example():
    start_time = datetime.now()
    x, y, a = (-538778789133922.0, -538778789133922.0, -2.6221616798653463e-203)
    config = python.Config()

    ui_Model = ui.Model(
        ui.Symbol("dt"),
        set(ui.symbols(["x", "y"])),
        set(ui.symbols(["a"])),
        {
            ui.Symbol("x"): ui.Symbol("x") * ui.Symbol("y") + ui.Symbol("x"),
            ui.Symbol("y"): "y + a * dt",
        },
    )
    ekf = python.ExtendedKalmanFilter(
        state_model=ui_Model,
        process_noise={ui.Symbol("a"): 1.0},
        sensor_models={"simple": {ui.Symbol("x"): ui.Symbol("x")}},
        sensor_noises={"simple": np.eye(1)},
        config=config,
    )

    control_vector = np.array([[a]])
    covariance = np.eye(2)
    state_vector = np.array([[x, y]]).transpose()

    if not np.isfinite(state_vector).all() or not np.isfinite(control_vector).all():
        # reject infinite inputs
        reject()
    if (np.abs(state_vector) > 1e100).any() or (np.abs(control_vector) > 1e100).any():
        # reject poorly sized inputs
        reject()

    next_state, next_cov = ekf.sensor_model(
        "simple", state=state_vector, covariance=covariance, sensor_reading=np.eye(1)
    )
    first_innovation = ekf.innovations["simple"]

    try:
        starting_central_probability = multivariate_normal(cov=covariance).pdf(
            np.zeros_like(state_vector).transpose()
        )
        ending_central_probability = multivariate_normal(cov=next_cov).pdf(
            np.zeros_like(state_vector).transpose()
        )
    except np.linalg.LinAlgError:
        print("starting cov")
        print(covariance)
        print("next_cov")
        print(next_cov)
        raise

    try:
        # more confident in state estimate after sensor update
        assert ending_central_probability > starting_central_probability
    except AssertionError:
        print("starting pdf")
        print(starting_central_probability)
        print("ending pdf")
        print(ending_central_probability)
        raise

    ekf.sensor_model(
        "simple", state=next_state, covariance=next_cov, sensor_reading=np.eye(1)
    )
    second_innovation = ekf.innovations["simple"]

    # state moves towards reading after a sensor update
    assert (
        np.linalg.norm(first_innovation) > np.linalg.norm(second_innovation)
        or np.linalg.norm(first_innovation) == 0.0
    )

    end_time = datetime.now()

    elapsed = end_time - start_time

    assert elapsed < timedelta(seconds=0.1)

File: py/test/formak/ExtendedKalmanFilter_test.cpp
#include <formak/testing/stats.h>
#include <gtest/gtest.h>
#include <rapidcheck/gtest.h>
#include <unit/simple-ekf.h>  // Generated

namespace integration {

namespace ekf_process_property_test {
using formak::testing::stats::MultivariateNormal;

RC_GTEST_PROP(CppModel, EKF_process_property, (double x, double y, double a)) {
  // def test_EKF_process_property(state_x, state_y, control_a):
  unit::ExtendedKalmanFilter ekf;
  double dt = 0.1;

  unit::Control control({a});
  unit::State state({x, y});
  unit::Covariance covariance;

  RC_PRE(std::isfinite(x) && std::isfinite(y) && std::isfinite(a));
  RC_PRE(std::abs(x) < 1e100 && std::abs(y) < 1e100 && std::abs(a) < 1e100);

  auto next = ekf.process_model(dt, {state, covariance}, control);

  RC_ASSERT(next.state.x() == x * y);
  RC_ASSERT(next.state.y() == y + a * dt);

  // try
  double starting_central_probability =
      MultivariateNormal(unit::State{}, covariance).pdf(unit::State{});
  double ending_central_probability =
      MultivariateNormal(unit::State{}, next.covariance).pdf(unit::State{});
  //     except np.linalg.LinAlgError:

  std::cout << "next.covariance" << std::endl;
  std::cout << next.covariance.data << std::endl;

  RC_ASSERT(ending_central_probability < starting_central_probability);
}

struct Options {
  double x;
  double y;
  double a;
};

class CppModelFailureCasesProcess
    : public ::testing::Test,
      public ::testing::WithParamInterface<Options> {};

TEST_P(CppModelFailureCasesProcess, RerunCases) {
  double a = GetParam().a;
  double x = GetParam().x;
  double y = GetParam().y;

  unit::ExtendedKalmanFilter ekf;
  double dt = 0.1;

  unit::Control control({a});
  unit::State state({x, y});
  unit::Covariance covariance;

  // Don't need pre for hand-inspected test cases
  // RC_PRE(std::isfinite(x) && std::isfinite(y) && std::isfinite(a));
  // RC_PRE(std::abs(x) < 1e100 && std::abs(y) < 1e100 && std::abs(a) < 1e100);

  auto next = ekf.process_model(dt, {state, covariance}, control);

  EXPECT_EQ(next.state.x(), x * y);
  EXPECT_EQ(next.state.y(), y + a * dt);

  // try
  double starting_central_probability =
      MultivariateNormal(unit::State{}, covariance).pdf(unit::State{});
  double ending_central_probability =
      MultivariateNormal(unit::State{}, next.covariance).pdf(unit::State{});
  //     except np.linalg.LinAlgError:

  std::cout << "next.covariance" << std::endl;
  std::cout << next.covariance.data << std::endl;

  EXPECT_LT(ending_central_probability, starting_central_probability);
}

INSTANTIATE_TEST_SUITE_P(PreviousFailureCases, CppModelFailureCasesProcess,
                         ::testing::Values(Options{0.0, 0.0, 0.0}));

}  // namespace ekf_process_property_test

namespace ekf_sensor_property_test {
using formak::testing::stats::MultivariateNormal;

RC_GTEST_PROP(CppModel, EKF_sensor_property, (double x, double y)) {
  // def test_EKF_sensor_property(x, y, a):
  unit::ExtendedKalmanFilter ekf;

  unit::State state({x, y});
  unit::Covariance covariance;

  RC_PRE(std::isfinite(x) && std::isfinite(y));
  RC_PRE(std::abs(x) < 1e100 && std::abs(y) < 1e100);

  double reading = 1.0;
  unit::SensorReading<(unit::SensorId::SIMPLE), unit::Simple> simple_reading{
      unit::Simple({reading})};
  auto next = ekf.sensor_model({state, covariance}, simple_reading);

  auto maybe_first_innovation =
      ekf.innovations<unit::SensorId::SIMPLE, unit::Simple>();
  RC_ASSERT(maybe_first_innovation.has_value());
  typename unit::Simple::InnovationT first_innovation =
      maybe_first_innovation.value();

  // try
  double starting_central_probability =
      MultivariateNormal(unit::State{}, covariance).pdf(unit::State{});
  double ending_central_probability =
      MultivariateNormal(unit::State{}, next.covariance).pdf(unit::State{});
  //     except np.linalg.LinAlgError:

  RC_ASSERT(ending_central_probability > starting_central_probability);

  // Run this to get a second innovation
  [[maybe_unused]] auto next2 = ekf.sensor_model(next, simple_reading);

  auto maybe_second_innovation =
      ekf.innovations<unit::SensorId::SIMPLE, unit::Simple>();
  RC_ASSERT(maybe_second_innovation.has_value());
  typename unit::Simple::InnovationT second_innovation =
      maybe_second_innovation.value();

  RC_ASSERT(first_innovation.norm() > second_innovation.norm() ||
            first_innovation.norm() == 0.0);
}

struct Options {
  double x;
  double y;
};

class CppModelFailureCasesSensor
    : public ::testing::Test,
      public ::testing::WithParamInterface<Options> {};

TEST_P(CppModelFailureCasesSensor, RerunCases) {
  double x = GetParam().x;
  double y = GetParam().y;

  // def test_EKF_sensor_property(x, y, a):
  unit::ExtendedKalmanFilter ekf;

  unit::State state({x, y});
  unit::Covariance covariance;

  // Don't need pre for hand-inspected test cases
  // RC_PRE(std::isfinite(x) && std::isfinite(y));
  // RC_PRE(std::abs(x) < 1e100 && std::abs(y) < 1e100);

  double reading = 1.0;
  unit::SensorReading<(unit::SensorId::SIMPLE), unit::Simple> simple_reading{
      unit::Simple({reading})};
  auto next = ekf.sensor_model({state, covariance}, simple_reading);

  auto maybe_first_innovation =
      ekf.innovations<unit::SensorId::SIMPLE, unit::Simple>();
  ASSERT_TRUE(maybe_first_innovation.has_value());
  typename unit::Simple::InnovationT first_innovation =
      maybe_first_innovation.value();

  // try
  double starting_central_probability =
      MultivariateNormal(unit::State{}, covariance).pdf(unit::State{});
  double ending_central_probability =
      MultivariateNormal(unit::State{}, next.covariance).pdf(unit::State{});
  //     except np.linalg.LinAlgError:

  RC_ASSERT(ending_central_probability > starting_central_probability);

  // Run this to get a second innovation
  [[maybe_unused]] auto next2 = ekf.sensor_model(next, simple_reading);

  auto maybe_second_innovation =
      ekf.innovations<unit::SensorId::SIMPLE, unit::Simple>();
  ASSERT_TRUE(maybe_second_innovation.has_value());
  typename unit::Simple::InnovationT second_innovation =
      maybe_second_innovation.value();

  EXPECT_TRUE(first_innovation.norm() > second_innovation.norm() ||
              first_innovation.norm() == 0.0);
}

INSTANTIATE_TEST_SUITE_P(PreviousFailureCases, CppModelFailureCasesSensor,
                         ::testing::Values(Options{0.0, 0.0}));

}  // namespace ekf_sensor_property_test

}  // namespace integration
